### DISCUSSION TITLE
TOTP support (Google Authenticator, Aegis, …) together with WebAuthn for two-factor authentication (2FA)

### DISCUSSION THREAD

**Discussion by rugk** [anonaddy/anonaddy#296] [Feature Requests]
It would be great if you could also support [TOTP](https://en.wikipedia.org/wiki/Time-based_one-time_password) in addition to U2F/WebAuthn (i.e. hardware keys) + backup tokens, you currently already have.

This would allow a better backup strategy/multiple options to use, especially if one does not have a hardware key (although that would likely be better).

## App support

Many apps are there which you can use to scan the QR code you get:
https://search.f-droid.org/?q=totp

Most famously Google Authenticator and Microsoft Authenticator, though both not the best ones, possibly.
[Aegis](https://f-droid.org/packages/com.beemdevelopment.aegis/) is e.g. a very good one.

**Comment by willbrowningme** [comment]
TOTP is already supported, however you can currently only have either TOTP or WebAuthn two factor enabled at any one time.

If you can't see TOTP you need to disable WebAuthn then you will be able to see it.

**Reply by rugk** [reply]
Well... there is no reason to implement it like this, i.e. so I assume this is a missing feature. As I said, TOTP is a better fallback to WebAuthn than these static backup codes, so well...

**Reply by willbrowningme** [reply]
I agree it would be nice to have both enabled at the same time and to be choose which to use, I'll see if I can add this in the near future.

**Reply by ghost** [reply]
hi, 
is there any update about this ? it will be a really good security feature if you add hardware keys as a 2fa method along side totp and we chose what to use everytime we try to login.

**Reply by willbrowningme** [reply]
No update at the moment I'm afraid, it would be good to see some code of someone who has already implemented this with Laravel so I can see how they achieved it.

**Reply by rugk** [reply]
I am not sure if this is Lavarel, but that is python is not it? Here you can see how the German community site FragDenStaat implemented it: https://github.com/okfde/froide/pull/480

The framework used is also linked there.

**Comment by uhlhosting** [comment]
@willbrowningme hi, is there a way to disable manually from cli the 2fa for a given user? It seems i got myself locked.

**Reply by willbrowningme** [reply]
You can run:

`php artisan tinker` from the web app directory.

Then find your user:

```php
$user = Username::firstWhere('username', 'johndoe')?->user;
```

Then if you have TOTP 2FA run:

```php
$user->two_factor_enabled = false;
$user->save();
```
If you use webAuthn run:

```php
$user->webauthnKeys()->delete();
```
